logic/src/commonMain/kotlin/com/wire/kalium/logic/feature/message/GetNotificationsUseCase.kt

```diff
@@ -0,0 +1,92 @@
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.take
+
+class GetNotificationsUseCase(
+    private val messageRepository: MessageRepository,
+    private val userRepository: UserRepository,
+    private val conversationRepository: ConversationRepository
+) {
+
+    suspend operator fun invoke(): Flow<List<DbNotificationConversation>> {
+        return messageRepository.getMessagesForNotification()
+            .flatMapMerge { dbMessages ->
+                val authorIds = dbMessages.map { it.senderUserId }.toSet()
+
+                flatMapFromIterable(authorIds) { userId -> userRepository.getKnownUser(userId) }
+                    .map { authors ->
+                        dbMessages.map { msg ->
+                            MessageWithSenderData(msg, authors.firstOrNull { it?.id == msg.senderUserId })
+                        }
+                    }
+            }
+            .map { it.groupBy { data -> data.message.conversationId } }
+            .flatMapMerge { map ->
+                flatMapFromIterable(map.keys) { conversationId ->
+                    conversationRepository.getConversationDetailsById(conversationId)
+                        .map { details ->
+                            val messagesWithUser = map[conversationId]
+                            val conversationName = details.conversation.name ?: ""
+
+                            val messages: List<DbNotificationMessage> = messagesWithUser?.map { msg ->
+                                val author = DbNotificationMessageAuthor(msg.user?.name ?: "", null)
+                                val time = msg.message.date
+
+                                when (msg.message.content) {
+                                    is MessageContent.Text -> {
+                                        val text = (msg.message.content as MessageContent.Text).value
+                                        DbNotificationMessage.Text(author, time, text)
+                                    }
+                                    else -> {
+                                        //TODO
+                                        DbNotificationMessage.Text(author, time, "Something not a text")
+                                    }
+                                }
+                            } ?: listOf()
+
+                            val isOneToOneConversation = details.conversation.type == Conversation.Type.ONE_ON_ONE
+
+                            DbNotificationConversation(
+                                id = conversationId,
+                                name = conversationName,
+                                messages = messages,
+                                isOneToOneConversation = isOneToOneConversation
+                            )
+                        }
+                }
+            }
+    }
+
+    private data class MessageWithSenderData(val message: Message, val user: OtherUser?)
+}
+
+private suspend fun <A, B> flatMapFromIterable(
```

Comment:
What do you think of moving this cool generic mapping function to a utility class, so we can reuse it later? Something like we have with `Either` or `SuspendableEitherScope`  for example with the `coFold` functions on functional package :D

Reply:
i was going to do that, but couldn't find a propper place, thanks :)